algorithms/src/snark/marlin/ahp/prover/round_functions/third.rs

```diff
@@ -21,249 +20,286 @@ use crate::{
+        alpha: &F,
+        eta_b: &F,
+        eta_c: &F,
+    ) -> Result<(DensePolynomial<F>, DensePolynomial<F>, ThirdMessage<F>), anyhow::Error> {
+        let num_instances = batch_combiners.values().map(|c| c.instance_combiners.len()).collect_vec();
+        let mut job_pool = ExecutionPool::with_capacity(num_instances.iter().sum::<usize>() * 3);
+        let max_variable_domain = &state.max_variable_domain;
+        let matrix_labels = ["a", "b", "c"];
+        let matrix_combiners = [F::one(), *eta_b, *eta_c];
+
+        for (((circuit, circuit_specific_state), batch_combiner), w_polys) in state
+            .circuit_specific_states
+            .iter_mut()
+            .zip(batch_combiners.values())
+            .zip(state.first_round_oracles.as_ref().unwrap().batches.values())
+        {
+            let circuit_combiner = batch_combiner.circuit_combiner;
+            let instance_combiners = &batch_combiner.instance_combiners;
+            let x_polys = &circuit_specific_state.x_polys;
+            let constraint_domain = &circuit_specific_state.constraint_domain;
+            let variable_domain = &circuit_specific_state.variable_domain;
+            let input_domain = &circuit_specific_state.input_domain;
+            let fft_precomputation = &circuit.fft_precomputation;
+            let ifft_precomputation = &circuit.ifft_precomputation;
+            let matrix_arith = [&circuit.a_arith, &circuit.b_arith, &circuit.c_arith];
+
+            let columns_to_non_zero_indices_time =
+                start_timer!(|| format!("Precomputing columns_to_non_zero_indices for circuit {}", circuit.id));
+            let mut c_i_to_k = BTreeMap::new(); // mapping of which column indices are used in which k non-zero entries
+            c_i_to_k.insert("a", non_zero_entries_per_column(&circuit.a, variable_domain, input_domain)?);
+            c_i_to_k.insert("b", non_zero_entries_per_column(&circuit.b, variable_domain, input_domain)?);
+            c_i_to_k.insert("c", non_zero_entries_per_column(&circuit.c, variable_domain, input_domain)?);
+            end_timer!(columns_to_non_zero_indices_time);
+
+            for (_j, (&instance_combiner, w_poly, x_poly)) in
+                itertools::izip!(instance_combiners, w_polys, x_polys).enumerate()
+            {
+                let z_time = start_timer!(move || format!("Compute z poly for circuit {} {}", circuit.id, _j));
+                let mut assignment = w_poly.0.polynomial().as_dense().unwrap().mul_by_vanishing_poly(*input_domain);
+                // Zip safety: `x_poly` is smaller than `z_poly`.
+                assignment.coeffs.iter_mut().zip(&x_poly.coeffs).for_each(|(z, x)| *z += x);
+                end_timer!(z_time);
+
+                for (label, matrix_combiner, arith) in itertools::izip!(matrix_labels, matrix_combiners, matrix_arith) {
+                    let lineval_assignment = assignment.clone(); // further below, PolyMultiplier does not want just a reference
```

Comment:
we can add an API for `PolyMultiplier` to take a reference too.

Reply:
Ha turns out it was already there: 5551f024c